lib/compress.js

```diff
@@ -1731,6 +1731,7 @@ merge(Compressor.prototype, {
                 var val = {};
                 for (var i = 0, len = this.properties.length; i < len; i++) {
                     var prop = this.properties[i];
+                    if (prop instanceof AST_Expansion) return this;
```

Comment:
Haven't got a PC around to test your patch, but wouldn't that suppress optimisation of `({a:b}).a`?

`expr.may_throw_on_access()` asks if `expr.foo` would be safe to eliminate; your patch propagates that from `({a:b}).foo` to `b.foo`, which seems a bit off.